pull.go

```diff
@@ -0,0 +1,435 @@
+		// Sort and limit manifests if a finite number is needed
+		if limit > 0 {
+			childrenHandler = images.LimitManifests(childrenHandler, rCtx.PlatformMatcher, limit)
+		}
+		if rCtx.Unpack {
+			unpacker = c.newUnpacker(rCtx, nil)
+			childrenHandler = unpacker.inform(childrenHandler)
+		}
+
+		handler = images.Handlers(append(rCtx.BaseHandlers,
+			remotes.FetchHandler(store, fetcher),
+			childrenHandler,
+		)...)
+	}
+	if unpacker != nil {
+		unpacker.start(ctx, desc)
+		// Make sure unpacker is stopped in all cases.
+		defer unpacker.stop(ctx, desc)
+	}
+
+	if err := images.Dispatch(ctx, handler, desc); err != nil {
+		return images.Image{}, err
+	}
+	if schema1Converter != nil {
+		desc, err = schema1Converter.Convert(ctx)
+		if err != nil {
+			return images.Image{}, err
+		}
+	}
+
+	if unpacker != nil {
+		if err := unpacker.stop(ctx, desc); err != nil {
+			return images.Image{}, err
+		}
+	}
+
+	img := images.Image{
+		Name:   name,
+		Target: desc,
+		Labels: rCtx.Labels,
+	}
+
+	is := c.ImageService()
+	for {
+		if created, err := is.Create(ctx, img); err != nil {
+			if !errdefs.IsAlreadyExists(err) {
+				return images.Image{}, err
+			}
+
+			updated, err := is.Update(ctx, img)
+			if err != nil {
+				// if image was removed, try create again
+				if errdefs.IsNotFound(err) {
+					continue
+				}
+				return images.Image{}, err
+			}
+
+			img = updated
+		} else {
+			img = created
+		}
+
+		return img, nil
+	}
+}
+
+type layerState struct {
+	layer      rootfs.Layer
+	downloaded bool
+	unpacked   bool
+	empty      bool
+}
+
+type unpackResult struct {
+	chain []digest.Digest
+	err   error
+}
+
+type unpacker struct {
+	stopped          bool
+	initCh           chan []rootfs.Layer
+	updateCh         chan *rootfs.Layer
+	resCh            chan unpackResult
+	c                *Client
+	snapshotter      string
+	platformMatcher  platforms.MatchComparer
+	schema1Converter *schema1.Converter
+}
+
+func (c *Client) newUnpacker(rCtx *RemoteContext, schema1Converter *schema1.Converter) *unpacker {
+	return &unpacker{
+		initCh:           make(chan []rootfs.Layer),
+		updateCh:         make(chan *rootfs.Layer, 4096),
+		resCh:            make(chan unpackResult),
+		c:                c,
+		snapshotter:      rCtx.Snapshotter,
+		platformMatcher:  rCtx.PlatformMatcher,
+		schema1Converter: schema1Converter,
+	}
+}
+
+func (u *unpacker) start(ctx context.Context, desc ocispec.Descriptor) {
+	go func() {
+		chain, err := func() ([]digest.Digest, error) {
+			var (
+				states []layerState
+				layers []rootfs.Layer
+				ok     bool
+			)
+			select {
+			case layers, ok = <-u.initCh:
+				if !ok {
+					return nil, errors.New("init channel closed")
+				}
+			case <-ctx.Done():
+				return nil, errors.New("context cancelled")
+			}
+			if u.schema1Converter == nil {
+				// Wait for diffIDs for non-schema1 image.
+				var diffIDs []rootfs.Layer
+				select {
+				case diffIDs, ok = <-u.initCh:
```

Comment:
it seems that it doesn't consider the manifest list(index) case, for example, pull `hello-world:latest` with --all-platform, if my understanding is correct

Reply:
`Pull` only supports one platform, see https://github.com/containerd/containerd/blob/master/client.go#L377.

I can add a comment to explain this.

Reply:
Done